pkg/core/native/management.go

```diff
@@ -89,6 +96,16 @@ func (m *Management) getContract(ic *interop.Context, args []stackitem.Item) sta
 
 // GetContract returns contract with given hash from given DAO.
 func (m *Management) GetContract(d dao.DAO, hash util.Uint160) (*state.Contract, error) {
+	m.RLock()
+	cs, ok := m.contracts[hash]
+	m.RUnlock()
+	if ok && cs != nil {
+		return cs, nil
+	}
+	return m.getContractFromDAO(d, hash)
```

Comment:
The trick here is to exactly _not_ go into DAO unless there was a change. We cache _everything_ and the cache can only be invalid for _some_ contracts that did change in some way. It's important as a NEP-17 transfer optimization, because any NEP-17 contract will make this lookup into the DB and 99,999% there will be nothing there for the `hash`. And we well know that just by looking at `m.contracts` where there is nothing for it.

What we can do is use `nil` value stored as a sign of creation/update/change. So for `!ok` you _always_ return `nil, ErrNotFound`, but for `ok` you either return the value got from `m.contracts` (is it's `!nil`) or you go into the DAO and return whatever it returns.

Technically, we'll also need to redesign our caching system, but that we'll do a bit later.

Reply:
This also means that you need to prime this cache during initialization, BTW.

Reply:
Fixed.